File: packages/zodiac-js/src/western/elements.ts
/**
 * Western Zodiac elements data module.
 */

import type { ZodiacElement } from '../models/zodiac-element';

export const WESTERN_ELEMENTS: readonly ZodiacElement[] = [
  { id: 'fire', name: 'Fire', system: 'western', members: ['aries', 'leo', 'sagittarius'] },
  { id: 'earth', name: 'Earth', system: 'western', members: ['taurus', 'virgo', 'capricorn'] },
  { id: 'air', name: 'Air', system: 'western', members: ['gemini', 'libra', 'aquarius'] },
  { id: 'water', name: 'Water', system: 'western', members: ['cancer', 'scorpio', 'pisces'] },
];

const elementMap = new Map<string, ZodiacElement>(
  WESTERN_ELEMENTS.map((e) => [e.id, e]),
);

/** Returns a Western element by ID. Throws if not found. */
export function getWesternElement(id: string): ZodiacElement {
  const el = elementMap.get(id);
  if (!el) {
    throw new Error(`Western element not found: "${id}"`);
  }
  return el;
}

/** Returns all 4 Western elements. */
export function getAllWesternElements(): readonly ZodiacElement[] {
  return WESTERN_ELEMENTS;
}

File: packages/zodiac-dart/lib/src/western/elements.dart
import '../models/zodiac_element.dart';

const List<ZodiacElement> westernElements = [
  ZodiacElement(
      id: 'fire',
      name: 'Fire',
      system: 'western',
      members: ['aries', 'leo', 'sagittarius']),
  ZodiacElement(
      id: 'earth',
      name: 'Earth',
      system: 'western',
      members: ['taurus', 'virgo', 'capricorn']),
  ZodiacElement(
      id: 'air',
      name: 'Air',
      system: 'western',
      members: ['gemini', 'libra', 'aquarius']),
  ZodiacElement(
      id: 'water',
      name: 'Water',
      system: 'western',
      members: ['cancer', 'scorpio', 'pisces']),
];

final Map<String, ZodiacElement> _elementMap = {
  for (var e in westernElements) e.id: e
};

ZodiacElement getWesternElement(String id) {
  final el = _elementMap[id];
  if (el == null) throw Exception('Western element not found: "$id"');
  return el;
}
